### ISSUE TITLE
Define _SOURCE feature macros in the configure script

### ISSUE THREAD

**Issue by omasanori** [ocaml/ocaml#12417]
Context: https://github.com/ocaml/ocaml/pull/12414#issuecomment-1646844277

Currently, we have three `_GNU_SOURCE` and one `_XOPEN_SOURCE` defined unconditionally in the runtime while they are actually needed for specific platforms. Especially, `_XOPEN_SOURCE` often hides some non-portable extensions, so defining it unconditionally leads to *less* declarations on some platforms.

In principle, the configure script should detect if a declaration is behind a feature flag in the target platform and define in such cases.

**Comment by xavierleroy** [comment]
The GNU libc documentation explicitly recommends to define feature macros at the beginning of C file that need them : https://www.gnu.org/software/libc/manual/html_node/Feature-Test-Macros.html .

Also, I vaguely remember a case where  defining`_GNU_SOURCE` would pick the wrong version of a library function, i.e. the pre-standard GNU interface, while we wanted to use the newer, standardized interface.  So, I would not want to define `_GNU_SOURCE` for all C source files.

In summary: I think the current situation is preferable to what is suggested in this issue report, so I'll take the liberty to close it.  Re-open if you disagree.

**Comment by omasanori** [comment]
No strong objection from my side. Thank you for your prompt response!

> Also, I vaguely remember a case where defining _GNU_SOURCE would pick the wrong version of a library function, i.e. the pre-standard GNU interface, while we wanted to use the newer, standardized interface. So, I would not want to define _GNU_SOURCE for all C source files.

Ah, yes, `strerror_r` is, for example.
